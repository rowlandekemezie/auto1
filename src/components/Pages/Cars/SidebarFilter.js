import React, { useState, useEffect } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import Button from "../../Button";
import FilterBox from "../../FilterBox";

import { renderOptions, buildQueryString } from "../../../utils";

const Wrapper = styled("div")`
  display: flex;
  flex-direction: column;
  margin-right: 24px;
  padding: 24px;
  border: 2px solid ${({ theme }) => theme.GRAY};
  height: 30%;
  margin-right: 0;

  @media (min-width: 768px) {
    width: 33.333%;
    margin-right: 24px;
  }
`;

const FilterButtonWrapper = styled("div")`
  align-self: flex-end;
  margin-top: 24px;
`;

export default function SidebarFilter({
  fetchAllColors,
  fetchAllManufacturers,
  manufacturers: { manufacturers },
  colors: { colors },
  history,
  fetchAllCars
}) {
  const [color, setColor] = useState({ value: "", label: "All car colors" });
  const [manufacturer, setManufacturer] = useState({
    value: "",
    label: "All manufacturers"
  });

  useEffect(() => {
    fetchAllManufacturers();
    fetchAllColors();
  }, []);

  const handleFilter = () => {
    const filters = {
      manufacturer: manufacturer.value,
      color: color.value,
      page: 1
    };
    const query = buildQueryString(filters);
    history.push(`/?${query}`);
    fetchAllCars(query);
  };

  return (
    <Wrapper>
      <FilterBox
        value={color}
        handleOnChange={e => setColor(e)}
        title="color"
        options={renderOptions(colors, "All car colors")}
      />
      <FilterBox
        value={manufacturer}
        handleOnChange={e => setManufacturer(e)}
        title="manufacturer"
        options={renderOptions(manufacturers, "All manufacturers")}
      />
      <FilterButtonWrapper>
        <Button label="Filter" onClick={handleFilter} />
      </FilterButtonWrapper>
    </Wrapper>
  );
}

SidebarFilter.propTypes = {
  fetchAllColors: PropTypes.func.isRequired,
  fetchAllManufacturers: PropTypes.func.isRequired,
  manufacturers: PropTypes.shape({
    manufacturers: PropTypes.arrayOf(PropTypes.string.isRequired)
  }).isRequired,
  colors: PropTypes.shape({
    colors: PropTypes.arrayOf(PropTypes.string.isRequired)
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }),
  fetchAllCars: PropTypes.func.isRequired
};
