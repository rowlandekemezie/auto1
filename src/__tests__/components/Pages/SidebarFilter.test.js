import React from "react";
import { shallow, mount } from "enzyme";
import { Router } from "react-router-dom";
import history from "history/createBrowserHistory";
import SidebarFilter from "../../../components/Pages/Cars/SidebarFilter";

describe("<SidebarFilter />", () => {
  const props = {
    fetchAllColors: jest.fn(),
    fetchAllCars: jest.fn(),
    fetchAllManufacturers: jest.fn(),
    manufacturers: { manufactures: ["BMW", "FORD"] },
    colors: { colors: ["Red", "Yellow"] }
  };

  it("should render component properly", () => {
    const wrapper = shallow(<SidebarFilter {...props} />);

    expect(wrapper.length).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it("should render with the required proptypes", () => {
    const wrapper = mount(<SidebarFilter {...props} />);

    expect(wrapper.length).toBe(1);
    expect(wrapper.find("FilterBox").length).toBe(2);
    expect(wrapper.find("button").text()).toBe("Filter");
    expect(wrapper.props()).toBeObject();
    expect(wrapper.props().colors).toBeObject(props.colors);
    expect(wrapper.props().manufacturers).toBeObject(props.manufacturers);
  });
});
