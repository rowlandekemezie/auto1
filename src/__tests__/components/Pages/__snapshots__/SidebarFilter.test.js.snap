// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarFilter /> should render component properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SidebarFilter
    colors={
      Object {
        "colors": Array [
          "Red",
          "Yellow",
        ],
      }
    }
    fetchAllCars={[MockFunction]}
    fetchAllColors={[MockFunction]}
    fetchAllManufacturers={[MockFunction]}
    manufacturers={
      Object {
        "manufactures": Array [
          "BMW",
          "FORD",
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FilterBox
          handleOnChange={[Function]}
          options={
            Array [
              Object {
                "label": "All car colors",
                "value": "",
              },
              Object {
                "label": "Red",
                "value": "Red",
              },
              Object {
                "label": "Yellow",
                "value": "Yellow",
              },
            ]
          }
          title="color"
          value={
            Object {
              "label": "All car colors",
              "value": "",
            }
          }
        />,
        <FilterBox
          handleOnChange={[Function]}
          options={
            Array [
              Object {
                "label": "All manufacturers",
                "value": "",
              },
            ]
          }
          title="manufacturer"
          value={
            Object {
              "label": "All manufacturers",
              "value": "",
            }
          }
        />,
        <ForwardRef>
          <_default
            label="Filter"
            onClick={[Function]}
          />
        </ForwardRef>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleOnChange": [Function],
          "options": Array [
            Object {
              "label": "All car colors",
              "value": "",
            },
            Object {
              "label": "Red",
              "value": "Red",
            },
            Object {
              "label": "Yellow",
              "value": "Yellow",
            },
          ],
          "title": "color",
          "value": Object {
            "label": "All car colors",
            "value": "",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleOnChange": [Function],
          "options": Array [
            Object {
              "label": "All manufacturers",
              "value": "",
            },
          ],
          "title": "manufacturer",
          "value": Object {
            "label": "All manufacturers",
            "value": "",
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <_default
            label="Filter"
            onClick={[Function]}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "label": "Filter",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-ifAKCX",
            "isStatic": true,
            "rules": Array [
              "
  align-self: flex-end;
  margin-top: 24px;
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-ifAKCX",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        },
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bxivhb",
        "isStatic": false,
        "rules": Array [
          "
  display: flex;
  flex-direction: column;
  margin-right: 24px;
  padding: 24px;
  border: 2px solid ",
          [Function],
          ";
  height: 30%;
  margin-right: 0;

  @media (min-width: 768px) {
    width: 33.333%;
    margin-right: 24px;
  }
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bxivhb",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FilterBox
            handleOnChange={[Function]}
            options={
              Array [
                Object {
                  "label": "All car colors",
                  "value": "",
                },
                Object {
                  "label": "Red",
                  "value": "Red",
                },
                Object {
                  "label": "Yellow",
                  "value": "Yellow",
                },
              ]
            }
            title="color"
            value={
              Object {
                "label": "All car colors",
                "value": "",
              }
            }
          />,
          <FilterBox
            handleOnChange={[Function]}
            options={
              Array [
                Object {
                  "label": "All manufacturers",
                  "value": "",
                },
              ]
            }
            title="manufacturer"
            value={
              Object {
                "label": "All manufacturers",
                "value": "",
              }
            }
          />,
          <ForwardRef>
            <_default
              label="Filter"
              onClick={[Function]}
            />
          </ForwardRef>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleOnChange": [Function],
            "options": Array [
              Object {
                "label": "All car colors",
                "value": "",
              },
              Object {
                "label": "Red",
                "value": "Red",
              },
              Object {
                "label": "Yellow",
                "value": "Yellow",
              },
            ],
            "title": "color",
            "value": Object {
              "label": "All car colors",
              "value": "",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleOnChange": [Function],
            "options": Array [
              Object {
                "label": "All manufacturers",
                "value": "",
              },
            ],
            "title": "manufacturer",
            "value": Object {
              "label": "All manufacturers",
              "value": "",
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <_default
              label="Filter"
              onClick={[Function]}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "label": "Filter",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-ifAKCX",
              "isStatic": true,
              "rules": Array [
                "
  align-self: flex-end;
  margin-top: 24px;
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-ifAKCX",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          },
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bxivhb",
          "isStatic": false,
          "rules": Array [
            "
  display: flex;
  flex-direction: column;
  margin-right: 24px;
  padding: 24px;
  border: 2px solid ",
            [Function],
            ";
  height: 30%;
  margin-right: 0;

  @media (min-width: 768px) {
    width: 33.333%;
    margin-right: 24px;
  }
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bxivhb",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
